### ISSUE TITLE
Auto-delete emails after a certain time period

### ISSUE THREAD

**Issue by sharat87** [maildev/maildev#365]
We are using a private maildev instance to catch all emails out of our staging environments. Its super convenient, works quite well, and doesn't hurt our domain's email rep at all.

However, we need to regularly go and delete old emails here, or it'll get too large too soon. If there was a way to configure maildev to delete emails, 4 hours or so after receiving them, that'd be perfect. The duration of course could be configurable, but that's the general idea.

If there's interest in this feature, I'm willing to contribute an implementation for this. I'll need some pointers though, since I'm new to maildev's codebase.

Thank you for your work on maildev!

**Comment by sharat87** [comment]
Hey @tlaukkanen, you are right, and I've ended up with a _similar_ workaround based on a cron job. I just hoped there could be just a CLI flag to just auto delete mails that are 24h old. If that doesn't sound like a good addition to maildev, I suppose this issue can be closed.

Thanks!

**Comment by sharat87** [comment]
Hey @tlaukkanen, also, just coming back to this, that curl command will actually delete all emails, including those that were received just moments ago. That's not really a practical solution for my case. Ideally, I'd like to delete emails that were a day old, or some such configurable time period. It's like truncating old log files. We want to keep the recent entries and purge old ones.

Can we do that with a query param or something perhaps?

**Comment by tlaukkanen** [comment]
@sharat87 unfortunately the _delete all_ seems to be only option for now. That's been ok case for our tests when we can cleanup all mails just before we start to run our test run.

If you check the [routes](https://github.com/maildev/maildev/blob/master/lib/routes.js#L50-L66) there are only two endpoints for deleting emails: all or by id.

But if you check the code closely it seems that each email is just a file so it shouldn't be a large effort to add a functionality so that it would only delete files that are i.e. older than X days/hours:
https://github.com/maildev/maildev/blob/96248f8c38bd269f541dd91e60ad560f57eb46a0/lib/mailserver.js#L121-L131

**Comment by tlaukkanen** [comment]
@sharat87 well checking the current pull requests there even seems to be a ready solution for this kind of auto cleanup so you could take this pull request to your own fork of maildev: [Remove emails, received more than some time ago](https://github.com/maildev/maildev/pull/312/files)

**Comment by sharat87** [comment]
I'm currently using the following Python script as a cron job to achieve this:

```python
import http.client
import datetime
from base64 import b64encode
import json


headers = {"Authorization": b"Basic " + b64encode(b"basic_username:basic_password")}
host = "maildevhost"


def del_mail(id):
    print("Deleting mail", id)
    con = http.client.HTTPConnection(host)
    con.request("DELETE", "/email/" + id, headers=headers)
    res = con.getresponse()
    data = res.read()
    if res.status != 200:
        print("Failed to delete mail with id %r, response is %r %r %r." % (id, res.status, res.reason, res.read()))


con = http.client.HTTPConnection(host)
con.request("GET", "/email", headers=headers)
res = con.getresponse()
print(res.status, res.reason)
mails = json.loads(res.read().decode("utf-8"))

for mail in mails:
    if datetime.datetime.fromisoformat(mail["time"][:-1]) + datetime.timedelta(hours=24) < datetime.datetime.utcnow():
        del_mail(mail["id"])

```

**Comment by mri1998** [comment]
The wget in the container cannot understand method DELETE, an easy and universal way to delete all email is running "reboot" in container, with "restart: always" in docker-compose.yml.

**Comment by jumpingstone** [comment]
I use the following command to clean emails very 5 seconds:
watch -n 5 "curl -s -X GET  http://localhost:1080/email |jq '.[]|.id'|xargs -I{} curl -s -X DELETE http://localhost:1080/email/{}"

**Comment by djfarrelly** [comment]
Consolidating into #127, which is the canonical issue for mailbox retention (it now covers both count-based "keep last N" and age-based "delete older than N" variants — your exact use case). There's no built-in age-based auto-deletion today; the only API affordance is delete-all / delete-by-id, hence the cron+curl workarounds in this thread.

There's also an open PR directly implementing this — #312 ("Remove emails, received more than some time ago") — worth following/reviewing if you need it soon. Closing as a duplicate of #127; please follow that issue (and PR #312).
